include/gabia/secure_socket.hpp

```diff
@@ -0,0 +1,302 @@
+        encrypt_pdu(pdu, sequence);
+        asio::write(next_layer(), asio::buffer(pdu), error);
+        if (error) {
+            return 0;
+        }
+
+        return asio::buffer_size(sequence);
+    }
+
+    template <typename ConstBufferSequence>
+    auto write_some(const ConstBufferSequence& sequence) {
+        beast::error_code error = {};
+        auto n = write_some(sequence, error);
+        Expects(!error);
+        return n;
+    }
+
+    template <typename MutableBufferSequence>
+    std::size_t read_some(const MutableBufferSequence& sequence,
+                          beast::error_code& error) {
+        error = {};
+        if (decrypted_buffer.size() > 0) {
+            // We have already decrypted data ready to be retrieved from the
+            // buffer
+            auto consumed =
+                boost::asio::buffer_copy(sequence, decrypted_buffer.data());
+            decrypted_buffer.consume(consumed);
+            return consumed;
+        }
+
+        boost::endian::little_int16_t payload_length = 0;
+        size_t transferred = 0;
+        while (transferred < sizeof(payload_length) + payload_length +
+                                 crypto::aead_auth_tag_size) {
+            auto mb = raw_buffer.prepare(max_payload_length);
+            transferred += next_layer().read_some(mb, error);
+            if (error) {
+                return 0;
+            }
+            raw_buffer.commit(transferred);
+            extract_pdu_length(payload_length);
+        }
+
+        auto success = decrypt_pdu(payload_length);
+        if (!success) {
+            // Forged/spoofed message
+            // TODO: use a more appropriate error code
+            error = asio::error::no_permission;
+            decrypted_buffer.consume(payload_length);
+            return 0;
+        }
+
+        auto consumed =
+            boost::asio::buffer_copy(sequence, decrypted_buffer.data());
+        decrypted_buffer.consume(consumed);
+        return consumed;
+    }
+
+    template <typename MutableBufferSequence>
+    std::size_t read_some(const MutableBufferSequence& sequence) {
+        beast::error_code error = {};
+        auto n = read_some(sequence, error);
+        Expects(!error);
+        return n;
+    }
+
+    template <typename MutableBufferSequence, typename Handler>
+    auto async_read_some(MutableBufferSequence sequence, Handler&& handler) {
+        beast::async_completion<Handler, void(beast::error_code, size_t)> init{
+            handler};
+        if (decrypted_buffer.size() > 0) {
+            auto consumed =
+                boost::asio::buffer_copy(sequence, decrypted_buffer.data());
+            decrypted_buffer.consume(consumed);
+            next_layer().get_io_service().dispatch([
+                consumed, handler = std::move(init.completion_handler)
+            ]() mutable {
+                beast::error_code error{};
+                handler(error, consumed);
+            });
```

Comment:
No, it wouldn't work right with a strand. From
http://www.boost.org/doc/libs/1_64_0/doc/html/boost_asio/reference/asio_handler_invoke.html

_When asynchronous operations are composed from other asynchronous operations, all intermediate handlers should be invoked using the same method as the final handler. This is required to ensure that user-defined objects are not accessed in a way that may violate the guarantees._

This means that when the Asio implementation invokes itself as a continuation of the asynchronous operation, your code will cause it to bypass the strand.

The hooks are going to change in Boost 1.66.0 (e.g. Net-TS). Instead of overloading `asio_handler_invoke` you will just have a member function `get_executor` which calls the next layer's `get_executor`. And for the allocation hooks they are replaced with a free function `get_associated_allocator` which returns an **Allocator** type.

>do you know why asio uses these weird allocation hooks rather than std::allocator?

I guess it made sense at the time (Asio was written in 2003?) Net-TS changes it to resemble standard allocators.